### ISSUE TITLE
Compile error with ref returns in latest roslyn package

### ISSUE THREAD

**Issue by kg** [dotnet/roslyn#33157]
Using the latest roslyn package from myget (3.1.0.1910509 | 6.92 MB | Tue, 05 Feb 2019 14:52:36 GMT) compiling mono from source fails:

```
/home/kate/Projects/mono/external/corert/src/System.Private.CoreLib/src/System/Array.cs(936,25): error CS1073: Unexpected token 'ref'
/home/kate/Projects/mono/external/corert/src/System.Private.CoreLib/src/System/Array.cs(937,24): error CS1073: Unexpected token 'ref'
```

I tested an older package and it works fine (3.1.0.1906016 | 6.89 MB | Thu, 10 Jan 2019 16:04:39 GMT). I checked and the correct langversion is being passed and csc is reporting support for 7.0, 7.1 and 7.2

The relevant section of the array.cs source:
```csharp
#if __MonoCS__
            ref T p = ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData());
            int i = index;
            int j = index + length - 1;
            while (i < j)
            {
                T temp = Unsafe.Add(ref p, i);
                Unsafe.Add(ref p, i) = Unsafe.Add(ref p, j);
                Unsafe.Add(ref p, j) = temp;
                i++;
                j--;
            }
#else
            ref T first = ref Unsafe.Add(ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData()), index);
            ref T last = ref Unsafe.Add(ref Unsafe.Add(ref first, length), -1);
            do
            {
                T temp = first;
                first = last;
                last = temp;
                first = ref Unsafe.Add(ref first, 1);
                last = ref Unsafe.Add(ref last, -1);
            } while (Unsafe.IsAddressLessThan(ref first, ref last));
#endif
        }
```
The errors are not on the first use of ref return, they're on the ```first = ...``` in the loop.

**Comment by marek-safar** [comment]
@kg could you point out to full C# file?

@jaredpar this is parser error so it should be easily reproducible and it looks like `#else` directive might confuse the tokenizer

**Comment by kg** [comment]
https://github.com/mono/corert/blob/master/src/System.Private.CoreLib/src/System/Array.cs#L936 Should be this file

**Comment by kg** [comment]
These are the three packages I downloaded and tested
![image](https://user-images.githubusercontent.com/198130/52374339-dd78c180-2a11-11e9-97a1-7e56bca2562c.png)
I'll test with your reduced test case today. Perhaps it has to do with the compile switches we pass to roslyn?
`
/home/kate/Projects/mono/runtime/mono-wrapper  /home/kate/Projects/mono/external/roslyn-binaries/Prototypes/DefaultInterfaceImplementation/csc.exe  /codepage:65001 /nologo /noconfig /deterministic -unsafe -nostdlib -nowarn:612,618,3001,3002,3003,1635 -d:INSIDE_CORLIB,MONO_CULTURE_DATA -d:LIBC -d:FEATURE_PAL,GENERICS_WORK,FEATURE_LIST_PREDICATES,FEATURE_SERIALIZATION,FEATURE_ENCODINGNLS,FEATURE_ASCII,FEATURE_LATIN1,FEATURE_UTF7,FEATURE_UTF32,MONO_HYBRID_ENCODING_SUPPORT,FEATURE_ASYNC_IO,NEW_EXPERIMENTAL_ASYNC_IO,FEATURE_UTF32,FEATURE_EXCEPTIONDISPATCHINFO,FEATURE_CORRUPTING_EXCEPTIONS,FEATURE_EXCEPTION_NOTIFICATIONS,FEATURE_STRONGNAME_MIGRATION,FEATURE_USE_LCID,FEATURE_FUSION,FEATURE_CRYPTO,FEATURE_X509_SECURESTRINGS,FEATURE_SYNCHRONIZATIONCONTEXT,FEATURE_SYNCHRONIZATIONCONTEXT_WAIT,HAS_CORLIB_CONTRACTS -d:FEATURE_MACL -d:FEATURE_REMOTING,MONO_COM,FEATURE_COMINTEROP,FEATURE_ROLE_BASED_SECURITY -d:MONO_FEATURE_THREAD_ABORT -d:MONO_FEATURE_THREAD_SUSPEND_RESUME -d:MONO_FEATURE_MULTIPLE_APPDOMAINS -d:BIT64,REGISTRY_ASSEMBLY  -d:NET_4_0 -d:NET_4_5 -d:MONO -d:WIN_PLATFORM -d:BOOTSTRAP_BASIC -nowarn:1699 -nostdlib -optimize /features:peverify-compat /langversion:latest  -d:FEATURE_PAL,GENERICS_WORK,FEATURE_LIST_PREDICATES,FEATURE_SERIALIZATION,FEATURE_ENCODINGNLS,FEATURE_ASCII,FEATURE_LATIN1,FEATURE_UTF7,FEATURE_UTF32,MONO_HYBRID_ENCODING_SUPPORT,FEATURE_ASYNC_IO,NEW_EXPERIMENTAL_ASYNC_IO,FEATURE_UTF32,FEATURE_EXCEPTIONDISPATCHINFO,FEATURE_CORRUPTING_EXCEPTIONS,FEATURE_EXCEPTION_NOTIFICATIONS,FEATURE_STRONGNAME_MIGRATION,FEATURE_USE_LCID,FEATURE_FUSION,FEATURE_CRYPTO,FEATURE_X509_SECURESTRINGS,FEATURE_SYNCHRONIZATIONCONTEXT,FEATURE_SYNCHRONIZATIONCONTEXT_WAIT,HAS_CORLIB_CONTRACTS -d:FEATURE_MACL -d:FEATURE_REMOTING,MONO_COM,FEATURE_COMINTEROP,FEATURE_ROLE_BASED_SECURITY -d:MONO_FEATURE_THREAD_ABORT -d:MONO_FEATURE_THREAD_SUSPEND_RESUME -d:MONO_FEATURE_MULTIPLE_APPDOMAINS -resource:resources/charinfo.nlp -resource:resources/collation.core.bin -resource:resources/collation.tailoring.bin -resource:resources/collation.cjkCHS.bin -resource:resources/collation.cjkCHT.bin -resource:resources/collation.cjkJA.bin -resource:resources/collation.cjkKO.bin -resource:resources/collation.cjkKOlv2.bin -resource:LinkerDescriptor/mscorlib.xml -runtimemetadataversion:v4.0.30319 -d:MONO_FEATURE_SRE -d:MONO_FEATURE_CONSOLE     -target:library -out:../../class/lib/build-linux/mscorlib.dll  @./../../build/deps/linux_build__corlib.dll.response
`

**Comment by kg** [comment]
Can confirm it compiles successfully with your test case, so it must have something to do with nostdlib?

**Comment by jaredpar** [comment]
@kg 

```
home/kate/Projects/mono/external/roslyn-binaries/Prototypes/DefaultInterfaceImplementation/csc.exe
```

Just checking. You downloaded the packages and then copied them over the `DefaultInterfaceImplementation` prototype compiler? Asking because that seemed an odd place to put csc and also I can realistically see that old prototype compiler having this particular parse error.

**Comment by kg** [comment]
I retested just now by unpacking the vsix in a fresh folder

**Comment by kg** [comment]
I think what happened here is something went wrong with our submodules and overwrote the new package, because I can't reproduce this anymore. False alarm, closing
